### DISCUSSION TITLE
Ownership-Borrowing :  Exercise 3

### DISCUSSION THREAD

**Discussion by CocDap** [openedu101/Rust-bootcamp-2024#13] [Q&A] [answered]
## Cho đoạn code dưới đây

```rust
fn main() {
    let mut count: u32 = 1;
    let mut num: u64 = 1;
    let mut primes: Vec<u64> = Vec::new();
    primes.push(2);

    while count < 10001 {
        num += 2;
        if vectorIsPrime(num, primes) {
            count += 1;
            primes.push(num);
        }
    }
}

fn vectorIsPrime(num: u64, p: Vec<u64>) -> bool {
    for i in p {
        if num > i && num % i != 0 {
            return false;
        }
    }

    true
}

```
## Nhiệm vụ
+ Giải thích lỗi 
+ Sửa lỗi

**Comment by ghost** [comment, marked as answer]
* Issue: Ở vòng lặp đầu tiên, vector `primes` đã bị moved vào hàm `vectorIsPrime` và bị drop sau khi thoát khỏi hàm. Do đó những vòng lặp tiếp theo biến `primes` không thể sử dụng lại được.
* Fix:
  - `vectorIsPrime(num, primes)` -> `vectorIsPrime(num, &primes)`
  - `fn vectorIsPrime(num: u64, p: Vec<u64>) -> bool` -> `fn vectorIsPrime(num: u64, p: &[u64]) -> bool`
  - `if num > i && num % i != 0` -> `if num > *i && num % *i != 0`

**Reply by CocDap** [reply]
Tuyệt vời
